specs/fault-proof.md

```diff
@@ -0,0 +1,412 @@
+
+It is up to the user to index the special pre-image values by this key scheme,
+as there is no way to revert it to the original commitment without knowing said commitment or value.
+
+#### Type `4-128`: reserved range
+
+Range start and and both inclusive.
+
+This range of key types is reserved for future usage by the core protocol.
+E.g. version changes, contract migrations, chain-data, additional core features, etc.
+
+`128` specifically (`1000 0000` in binary) is reserved for key-type length-extension
+(reducing the content part to `30` or less key bytes), if the need arises.
+
+#### Type `129-255`: application usage
+
+This range of key types may be used by forks or customized versions of the fault proof protocol.
+
+### Bootstrapping
+
+Initial inputs are deterministic, but not necessarily singular or global:
+there may be multiple different disputes at the same time, each with its own disputed claims and L1 context.
+
+To bootstrap, the program requests the initial inputs from the VM, using pre-image key type `1`.
+
+The VM is aware of the external context, and maps requested pre-image keys based on their type, i.e.
+a local lookup for type `1`, or global one for `2`, and optionally support other key-types.
+
+### Hinting
+
+There is one more form of optional communication between client and server: pre-image hinting.
+Hinting is optional, and *is a no-op* in a L1 VM implementation.
+
+The hint itself comes at very low cost onchain: the hint can be a single `write` sys-call,
+which is instant as the memory to write as hint does not actually need to be loaded as part of the onchain proof.
+
+Hinting allows the program, when generating a proof offchain,
+to instruct the VM what data it is interested in.
+
+The VM can choose to execute the requested hint at any time: either locally (for standard requests),
+or in a modular form by redirecting the hint to tooling that may come with the VM program.
+
+Hints do not have to be executed directly: they may first just be logged to show the intents of the program,
+and the latest hint may be buffered for lazy execution, or dropped entirely when in read-only mode (like onchain).
+
+When the pre-image oracle serves a request, and the request cannot be served from an existing collection of pre-images
+(e.g. a local pre-image cache) then the VM can execute the hint to retrieve the missing pre-image(s).
+It is the responsiblity of the program to provide sufficient hinting for every pre-image request.
+Some hints may have to be repeated: the VM only has to execute the last hint when handling a missing pre-image.
+
+Note that hints may produce multiple pre-images:
+e.g. a hint for an ethereum block with transaction list may prepare pre-images for the header,
+each of the transactions, and the intermediate merkle-nodes that form the transactions-list Merkle Patricia Trie.
+
+Hinting is implemented with a minimal wire-protocol over a blocking one-way stream:
```

Comment:
The neat thing about the hinting concept is that we move the error handling from the program to the VM (which can pass it back to some pre-image fetching tool that comes separately with the program).

This makes it so that the program code-path that runs **onchain is the exact same as the one that runs offchain** during pre-image fetching, avoiding any surprises. The only difference would be the compiler target. But no differences in program logic.

And outside of the program the pre-image fetching can be deferred, cached, retried, etc. all without complicating the program logic or design.

Reply:
So does this mean that the offchain program is going to have a separate process & then do the read/writes? I still think that you need to intercept the syscalls somewhere & that might be different (though that is a smaller diff than the pre-image oracle).

My thinking so far is that the error handling is on the other side of an interface (so everything is still in memory + in process) rather than out of process.

Reply:
Both will be possible, the separation is optional, we can bind them together the other way for minimal tests as well. In the full production mode however I'd like the code-path during prefetching to be unified with the onchain execution as much as possible, and it's relatively easy to test still (use some pipes, or a sub-process if really unified all the way), and a lot easier than testing what unified execution with memory mapping would look like.

Reply:
Actually, I think if you specify it this way with an ABI you can run the same binary on chain as off chain. That maybe requires some shims off-chain (or implementing a very simple linux ABI on chain), but that's kinda cool.